python/cudf/cudf/pandas/fast_slow_proxy.py

```diff
@@ -1011,7 +1018,37 @@ def _transform_arg(
         # use __reduce_ex__ instead...
         if type(arg) is tuple:
             # Must come first to avoid infinite recursion
-            return tuple(_transform_arg(a, attribute_name, seen) for a in arg)
+            if (
+                len(arg) > 0
+                and isinstance(arg[0], _MethodProxy)
```

Comment:
We only want to follow this branch when we call `DataFrame.__setitem__` and the underlying wrapped object is a `pd.DataFrame` (so we avoid the DtoH transfer), right? Would a proxy object that wraps `cudf.DataFrame` also follow this code path?

Reply:
Yes and yes. This is currently written in a way we could easily add to the dict map above for any other function and parameter.

Reply:
Can you lmk if I'm misunderstanding? If a wrapped cudf.DataFrame goes through this code path, then it will incorrectly call "_fsproxy_slow" when it should have stayed "fast". Eg. where df is a wrapped `cudf.DataFrame`
`df.__setitem__("A", pd.Index([4,5,6]))`

Reply:
`df.__setitem__("A", pd.Index([4,5,6]))` will not force `df` into slow path. It will only for `"A"`(at 0th index) in slow path which is also `"A"`, now lets see this example:
`df.__setitem__(pd.Index(["a", "b"]), pd.Index([4,5,6]))`, this will force `pd.Index(["a", "b"])` to slow path with is a true pandas index rather than passing a cudf Index.

Reply:
I documented a bit more thoroughly here: https://github.com/rapidsai/cudf/pull/17222/commits/04ec5b7139eac2db8e7b87b03a5fec104a285808

Reply:
Thanks for adding that! I have a question about this example:
```
df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))
```
So `df` can either be a wrapped a pandas or cudf dataframe.
1. `df` is a wrapped a pandas dataframe. 
In this case, `pd.Index(["a", "b"])` is forced to take the slow path. This makes sense to because if we keep it a fast-slow proxy object, then it will be converted to `cudf.Index(["a", "b"])` and then cause `df.__setitem__(cudf.Index(["a", "b"]), ...)` to fail because you can't pass a GPU object to pandas like this. Because this would fail, we'd trigger a DtoH transfer. This data transfer is the main contributing factor to the slow down. This case makes sense to me.
2.`df` is a wrapped a cudf dataframe. 
In this case, `pd.Index(["a", "b"])` is also forced to take the slow path. Is this okay because `__setitem__` also fails in this case? Trying this locally, I get
```
TypeError: Index object is not iterable. Consider using `.to_arrow()`, `.to_pandas()` or `.values_host` if you wish to iterate over the values.
```

Reply:
Were you arriving at error 2 via public api or mocking the api? I think error 2 isn't possible via public apis.

Reply:
For the following case:

df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

It will be converted to this and tried with cudf:

df.__setitem__(pd.Index(["a", "b"]), cudf.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

And then if the above fails the following will be tried on pandas:

df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

Reply:
No it doesn't fail via the public API, I used `CUDF_PANDAS_FALLBACK_DEBUGGING=True`

Reply:
It turns out (2.) doesn't work even w/o `cudf.pandas`

```
In [1]: import cudf

In [2]: import pandas as pd

In [3]: df = cudf.DataFrame()

In [4]: df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

ValueError: Data must be 1-dimensional

In [5]: df = pd.DataFrame()

In [6]: df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

In [7]: df
Out[7]: 
   a   b
0  4  10
1  5  11
2  6  12
```

Reply:
Yep there's fallback in this case anyway.
```
In [1]: %load_ext cudf.pandas

In [2]: import pandas as pd

In [3]: df = pd.DataFrame()

In [4]: type(df._fsproxy_wrapped)
Out[4]: cudf.core.dataframe.DataFrame

In [5]: df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

In [6]: type(df._fsproxy_wrapped)
Out[6]: pandas.core.frame.DataFrame
```

Reply:
This what I get with your changes. (And running with `CUDF_PACDAS_DEBUGGING=True` does not fail). This looks correct to me
```
In [1]: %load_ext cudf.pandas

In [2]: import pandas as pd

In [3]: df = pd.DataFrame()

In [4]: type(df._fsproxy_wrapped)
Out[4]: cudf.core.dataframe.DataFrame

In [5]: df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

In [6]: type(df._fsproxy_wrapped)
Out[6]: cudf.core.dataframe.DataFrame
```